cairo-contracts/src/tests/mocks/transfer_mock.cairo

```diff
@@ -0,0 +1,142 @@
+    impl TransferInternalImpl = ICS20TransferComponent::TransferInternalImpl<ContractState>;
+
+    #[storage]
+    struct Storage {
+        #[substorage(v0)]
+        transferrable: TransferrableComponent::Storage,
+        #[substorage(v0)]
+        transfer: ICS20TransferComponent::Storage,
+    }
+
+    #[event]
+    #[derive(Debug, Drop, starknet::Event)]
+    pub enum Event {
+        #[flat]
+        TransferrableEvent: TransferrableComponent::Event,
+        #[flat]
+        ICS20TransferEvent: ICS20TransferComponent::Event,
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, erc20_class_hash: ClassHash) {
+        self.transferrable.initializer();
+        self.transfer.initializer(erc20_class_hash);
+    }
+
+    #[abi(embed_v0)]
+    impl TransferValidateImpl of ITransferValidate<ContractState> {
+        fn escrow_validate(
+            self: @ContractState,
+            from_account: ContractAddress,
+            port_id: PortId,
+            channel_id: ChannelId,
+            denom: ERC20Token,
+            amount: u256,
+            memo: Memo,
+        ) {
+            self.transfer.escrow_validate(from_account, port_id, channel_id, denom, amount, memo,);
+        }
+
+        fn unescrow_validate(
+            self: @ContractState,
+            to_account: ContractAddress,
+            port_id: PortId,
+            channel_id: ChannelId,
+            denom: ERC20Token,
+            amount: u256,
+        ) {
+            self.transfer.unescrow_validate(to_account, port_id, channel_id, denom, amount,);
+        }
+
+        fn mint_validate(
+            self: @ContractState, account: ContractAddress, denom: PrefixedDenom, amount: u256,
+        ) {
+            self.transfer.mint_validate(account, denom, amount);
+        }
+
+        fn burn_validate(
+            self: @ContractState,
+            account: ContractAddress,
+            denom: PrefixedDenom,
+            amount: u256,
+            memo: Memo,
+        ) {
+            self.transfer.burn_validate(account, denom, amount, memo);
+        }
+    }
+
+    #[abi(embed_v0)]
+    impl TransferExecuteImpl of ITransferExecute<ContractState> {
+        fn escrow_execute(
+            ref self: ContractState,
+            from_account: ContractAddress,
+            denom: ERC20Token,
+            amount: u256,
+            memo: Memo,
+        ) {
+            self.transfer.escrow_execute(from_account, denom, amount, memo);
+        }
+
+        fn unescrow_execute(
+            ref self: ContractState,
+            to_account: ContractAddress,
+            port_id: PortId,
+            channel_id: ChannelId,
+            denom: ERC20Token,
+            amount: u256,
+        ) {
+            self.transfer.unescrow_execute(to_account, port_id, channel_id, denom, amount);
+        }
+
+        fn mint_execute(
+            ref self: ContractState, account: ContractAddress, denom: PrefixedDenom, amount: u256,
```

Comment:
> I also don't really understand the use of PrefixedDenom::trace_path. My understanding is that we only need it to figure out whether we need to unescrow or mint.

Whether to unescrow or mint is decided based on the [denom's base type](https://github.com/informalsystems/ibc-starknet/blob/0e48bd97f60bd3fd0acc80c8ec576b64b2b62bb9/cairo-contracts/src/apps/transfer/component.cairo#L193). If the denom is of the `Denom::Hosted` type, [it's prefixed by a `trace_path`](https://github.com/informalsystems/ibc-starknet/blob/0e48bd97f60bd3fd0acc80c8ec576b64b2b62bb9/cairo-contracts/src/apps/transfer/component.cairo#L207).

This prefix is necessary for tracking IBC-created tokens. Used [here](https://github.com/informalsystems/ibc-starknet/blob/0e48bd97f60bd3fd0acc80c8ec576b64b2b62bb9/cairo-contracts/src/apps/transfer/component.cairo#L337) during key computation. Later on, if someone wants to send e.g. `uatom` back to its source, the `burn_execute` needs [to see if](https://github.com/informalsystems/ibc-starknet/blob/0e48bd97f60bd3fd0acc80c8ec576b64b2b62bb9/cairo-contracts/src/apps/transfer/component.cairo#L295) this `uatom` really came from the specific port, channel_Id pair specified in the receive packet. Basically taking into account multihop scenarios. That’s why prefix's included it in the key.

> I guess we can just treat the entire denom as string in the case of mint?

To make the `mint_execute` work with `ByteArray`, it actually doesn’t fit with the current impls. As we'll need to first convert the `PrefixedDenom` into a `ByteArray` before passing it to `mint_execute`. This adds an extra step since, after that, we'll need to parse it again to retrieve the base denom [if the token should be created](https://github.com/informalsystems/ibc-starknet/blob/0e48bd97f60bd3fd0acc80c8ec576b64b2b62bb9/cairo-contracts/src/apps/transfer/component.cairo#L343-L345).

Let me know if anything isn't clear. We can hop on a call if needed.

Reply:
Ok then. Let's merge this first and figure out how to improve the API later.